### PR TITLE
feat(typescript): add tutorial for writing policies with typescript/javascript

### PR DESCRIPTION
**Pull request by esosaoh**
## Description

Fix https://github.com/kubewarden/policy-sdk-js/issues/190

## Test

To test this pull request, you can run the following commands:

### CODE REVIEW

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+This test ensures the policy rejects Pods when their hostname is in the deny list.
+
+### Test 2: Accept allowed hostnames
+
+```bash
+@test "accept because hostname is not on the deny list" {
+  run kwctl run annotated-policy.wasm -r test_data/pod_with_hostname.json --settings-json '{"denied_hostnames": ["forbidden-host"]}'
+  
+  # this prints the output when one the checks below fails
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:142]:
```suggestion
  # this prints the output when one of the checks below fails
```

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+This test verifies the policy accepts Pods when their hostname is not in the deny list.
+
+### Test 3: Accept when no settings provided
+
+```bash
+@test "accept because the deny list is empty" {
+  run kwctl run annotated-policy.wasm -r test_data/pod_with_hostname.json
+  
+  # this prints the output when one the checks below fails
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:159]:
```suggestion
  # this prints the output when one of the checks below fails
```

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+This test verifies that Pods without hostnames are accepted regardless of the deny list.
+
+### Test 5: Accept non-Pod resources
+
+```bash
+@test "accept non-pod resources" {
+  run kwctl run annotated-policy.wasm -r test_data/deployment.json --settings-json '{"denied_hostnames": ["forbidden-host"]}'
+  
+  # this prints the output when one the checks below fails
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:193]:
```suggestion
  # this prints the output when one of the checks below fails
```

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+### Test 6: Settings validation
+
+You can also add tests to verify settings validation works correctly:
+
+```bash
+@test "accept valid settings" {
+  run kwctl run annotated-policy.wasm -r test_data/pod.json --settings-json '{"denied_hostnames": ["host1", "host2"]}'
+  
+  # this prints the output when one the checks below fails
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:216]:
```suggestion
  # this prints the output when one of the checks below fails
```

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+}
+```
+
+### Test 7: Edge cases
+
+```bash
+@test "reject with multiple denied hostnames" {
+  run kwctl run annotated-policy.wasm -r test_data/pod_with_hostname.json --settings-json '{"denied_hostnames": ["bad-host", "test-hostname", "forbidden-host"]}'
+  
+  # this prints the output when one the checks below fails
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:231]:
```suggestion
  # this prints the output when one of the checks below fails
```

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+```console
+make e2e-tests
+```
+
+This produces output like:
+
+```console
+bats e2e.bats
+e2e.bats
+ ✓ reject because hostname is on deny list
```
Comment (dariavladykina) [inline docs/tutorials/typescript/05-e2e-tests.md:267]:
```suggestion
 ✓ reject because hostname is on the deny list
```

File: docs/tutorials/typescript/02-scaffold.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+npx webpack --config webpack.config.cjs
+asset bundled.js 5.52 KiB [compared for emit] [minimized] (name: main)
+asset types.d.ts 430 bytes [compared for emit]
+asset index.d.ts 11 bytes [compared for emit]
+./src/index.ts 3.84 KiB [built] [code generated]
+./node_modules/kubewarden-policy-sdk/dist/bundle.js 3.85 KiB [built] [code generated]
+webpack 5.101.3 compiled successfully in 1909 ms
+bats e2e.bats
+e2e.bats
+ ✓ reject because hostname is on deny list
```
Comment (dariavladykina) [inline docs/tutorials/typescript/02-scaffold.md:114]:
```suggestion
 ✓ reject because hostname is on the deny list
```

File: docs/tutorials/typescript/01-intro-typescript.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+> Javy takes your JavaScript code and executes it in a WebAssembly context. 
+>
+> It features an embedded QuickJS engine, compiled to WebAssembly, that can execute JavaScript. 
+>
+> The project provides both a CLI and a set of APIs for embedding and customizing the behavior when running JavaScript in WebAssembly.
+
+The Kubewarden project currently uses Javy for these reasons:
+
+- Mature JavaScript engine (QuickJS) compiled to WebAssembly
+- Support for [waPC](https://wapc.io) through custom host functions
```
Comment (flavio) [inline docs/tutorials/typescript/01-intro-typescript.md:32]: This is not correct, we're not using waPC as part of the JS SDK. The policy we're writing is targeting our WASI interface

File: docs/tutorials/typescript/01-intro-typescript.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+## Getting the toolchain
+
+The easiest way to get the complete toolchain is by using the Kubewarden JavaScript SDK, which includes the Javy compilation plugin:
+
+```bash
+npm install kubewarden-policy-sdk
+```
+
+The Javy plugin binary is automatically included and can be found at:
```
Comment (flavio) [inline docs/tutorials/typescript/01-intro-typescript.md:74]: Sorry, unrelated with the docs... Does it mean the binary plugin is shipped via npm? If so, how do we deal with operating systems and architectures?

File: docs/tutorials/typescript/03-policy-settings.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+keywords: [kubewarden, kubernetes, defining policy settings, TypeScript]
+doc-type: [tutorial]
+doc-topic: [kubewarden, writing-policies, typescript, defining-policy-settings]
+doc-persona: [kubewarden-policy-developer]
+---
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/policy-settings"/>
+</head>
+
```
Comment (flavio) [inline docs/tutorials/typescript/03-policy-settings.md:15]: I think we should add a big fat warning tooltip that reminds developers they must not print logging information to the STDOUT, otherwise the policy will be broken.

Explain they have to either write to the STDERR or use the logging facility from our SDK (do we have it?)

File: docs/tutorials/typescript/04-validation.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+keywords: [kubewarden, kubernetes, writing policies, typescript]
+doc-type: [tutorial]
+doc-topic: [kubewarden, writing-policies, typescript, validation-logic]
+doc-persona: [kubewarden-policy-developer]
+---
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/validation"/>
+</head>
+
```
Comment (flavio) [inline docs/tutorials/typescript/04-validation.md:15]: We should put the very same warning box we put on the policy settings page, the one about not writing to STDOUT

File: docs/tutorials/typescript/04-validation.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+This function extracts the hostname from a Pod resource:
+
+```typescript
+/**
+ * Extracts the hostname from a Pod resource.
+ *
+ * @param {KubernetesResource} resource - The Kubernetes resource.
+ * @returns {string | undefined} The hostname if set, otherwise undefined.
+ */
+function getPodHostname(resource: KubernetesResource): string | undefined {
```
Comment (flavio) [inline docs/tutorials/typescript/04-validation.md:134]: why don't we use the Pod type instead?

File: docs/tutorials/typescript/05-e2e-tests.md [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+## Basic test cases
+
+The template already provides several basic tests. Here are the existing tests with explanations:
+
+### Test 1: Reject denied hostnames
+
+```bash
+@test "reject because hostname is on deny list" {
+  run kwctl run annotated-policy.wasm -r test_data/pod_with_hostname.json --settings-json '{"denied_hostnames": ["forbidden-host", "test-hostname"]}'
```
Comment (flavio) [inline docs/tutorials/typescript/05-e2e-tests.md:122]: Are you sure this works? `kwctl run` takes a parameter a Kubernetes `AdmissionRequest` object via its `-r` flag. What you're passing is a Kubernetes object definition.

You should either use a fixture that has an `AdmissionRequest` defined inside of it, or you should use `kwctl scaffold admission-request ...` to convert that Pod definition into an `AdmissionRequest` to be fed into `kwctl run`.

**Review by flavio — CHANGES_REQUESTED** [review]

**Review by jhkrug — CHANGES_REQUESTED** [review]
A minor few suggestions for consideration. From a few sources. I'll have another run through when they have been incorporated. Thank you. Looking good.

**Comment by jhkrug** [comment]
Hi @esosaoh I think it needs to be deleted from the old location too. Thanks.

File: docs/tutorials/typescript/01-intro-typescript.md (hunk 9 above)
Comment (flavio) [reply docs/tutorials/typescript/01-intro-typescript.md:82]: My fault, the plugin is actually a `.wasm` module, hence we don't need to build it for other OS/architectures.

**Review by jhkrug — CHANGES_REQUESTED** [review]
Two things for the canonical links. Also a link needs fixing, looking at the failed build log for Netlify. Otherwise looking good.

**Review by jhkrug — CHANGES_REQUESTED** [review]
One more link.

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
+---
+sidebar_label: Writing policies in TypeScript/JavaScript
+sidebar_position: 60
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:3]: Put into correct order.
```suggestion
sidebar_position: 100
```

File: docs/tutorials/writing-policies/typescript/02-scaffold.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+webpack 5.101.3 compiled successfully in 2280 ms
+npm install
+
+up to date, audited 400 packages in 2s
+
+58 packages are looking for funding
+  run `npm fund` for details
+
+found 0 vulnerabilities
+```
```
[the remark this answers is not in the document]
Comment (esosaoh) [reply docs/tutorials/writing-policies/typescript/02-scaffold.md:101]: i just tried it and got the same thing, i think we should keep it in the tutorial as just `make all` and make the fix in the template to run npm install first. something like:

```make
dist/bundled.js: src/index.ts package.json webpack.config.cjs
	@mkdir -p dist
	npm install
	npm run build
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+## Getting the toolchain
+
+The easiest way to get the complete toolchain is by using the Kubewarden JavaScript SDK, which includes the Javy compilation plugin:
+
+```bash
+npm install kubewarden-policy-sdk
+```
+
+The Javy plugin binary is automatically included and can be found at:
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:82]: Active voice is better.
```suggestion
The Javy plugin binary is automatically included and you can find it at:
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+```
+node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden
+```
+
+## Tutorial prerequisites
+
+During this tutorial you need these tools on your development machine:
+
+- **Node.js**: Version 18 or higher with npm for dependency management
+- [**bats**](https://github.com/bats-core/bats-core): Used to write the tests and automate their execution
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:93]: Consistency with kwctl, coming next. Both commands. Not that important. But consistent.
```suggestion
- [**`bats`**](https://github.com/bats-core/bats-core): Used to write the tests and automate their execution
```

File: docs/tutorials/writing-policies/typescript/03-policy-settings.md [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/policy-settings"/>
+</head>
+
+:::danger Critical: Do not write logging information to STDOUT
+
+Writing to STDOUT will break your policy. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.
+
+:::
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/03-policy-settings.md:20]:
```suggestion
:::danger Critical: Don't write logging information to STDOUT

Writing to STDOUT breaks your policy. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.

:::
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
A few more style suggestions. Thanks.

**Comment by viccuad** [comment]
Reopening, merging the linked PR closed this PR.

**Comment by esosaoh** [comment]
done! @jhkrug

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+These TypeScript/JavaScript libraries are useful when writing a Kubewarden policy:
+
+- [Kubewarden JavaScript SDK](https://github.com/kubewarden/policy-sdk-js): Provides structures and functions reducing the amount of code necessary. It also provides test helpers and access to all host capabilities.
+- [Kubernetes TypeScript types](https://github.com/silverlyra/kubernetes-types): Provides TypeScript definitions for all Kubernetes resources, enabling type-safe policy development.
+
+The Kubewarden project provides a [template JavaScript/TypeScript policy project](https://github.com/kubewarden/js-policy-template) you can use to create Kubewarden policies.
+
+## Getting the toolchain
+
+The easiest way to get the complete toolchain is by using the Kubewarden JavaScript SDK, which includes the Javy compilation plugin:
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:76]:
```suggestion
The easiest way to get the complete toolchain is by using the Kubewarden JavaScript SDK, which includes the Javy compilation plug-in:
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+```bash
+npm install kubewarden-policy-sdk
+```
+
+The Javy plugin binary is automatically included and you can find it at:
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:82]:
```suggestion
The Javy plug-in binary is automatically included and you can find it at:
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+## Javy limitations
+
+Javy runs JavaScript in a sandboxed WebAssembly environment with certain constraints:
+
+- **WASI environment only**: Access limited to stdin/stdout/stderr and explicitly provided host capabilities
+- **No Node.js APIs**: Standard Node.js modules like `fs`, `http`, or `crypto` aren't available
+- **Limited standard library**: Only core JavaScript features and explicitly enabled APIs are accessible
+- **Single-threaded execution**: No support for Web Workers or multi-threading
+- **STDOUT restrictions**: Writing to STDOUT will break your policy - use STDERR for logging instead
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:51]:
```suggestion
- **WASI environment only**: Access limited to stdin/stdout/stderr and explicitly provided host capabilities.
- **No Node.js APIs**: Standard Node.js modules like `fs`, `http`, or `crypto` aren't available.
- **Limited standard library**: Only core JavaScript features and explicitly enabled APIs are accessible.
- **Single-threaded execution**: No support for Web Workers or multi-threading.
- **STDOUT restrictions**: Writing to STDOUT breaks policies - use STDERR for logging instead.
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+> It features an embedded QuickJS engine compiled to WebAssembly that can execute JavaScript. 
+>
+> The project provides both a CLI and a set of APIs for embedding and customizing the behavior when running JavaScript in WebAssembly.
+
+The Kubewarden project currently uses Javy for these reasons:
+
+- Mature JavaScript engine (QuickJS) compiled to WebAssembly
+- Support for [WASI interface](../wasi/01-intro-wasi.md) through custom host functions
+- Smaller binary sizes compared to other JavaScript-to-WebAssembly solutions
+- Active development and maintenance by the Bytecode Alliance
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:41]:
```suggestion
- Mature JavaScript engine (QuickJS) compiled to WebAssembly.
- Support for [WASI interface](../wasi/01-intro-wasi.md) through custom host functions.
- Smaller binary sizes compared to other JavaScript-to-WebAssembly solutions.
- Active development and maintenance by the Bytecode Alliance.
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+Javy runs JavaScript in a sandboxed WebAssembly environment with certain constraints:
+
+- **WASI environment only**: Access limited to stdin/stdout/stderr and explicitly provided host capabilities
+- **No Node.js APIs**: Standard Node.js modules like `fs`, `http`, or `crypto` aren't available
+- **Limited standard library**: Only core JavaScript features and explicitly enabled APIs are accessible
+- **Single-threaded execution**: No support for Web Workers or multi-threading
+- **STDOUT restrictions**: Writing to STDOUT will break your policy - use STDERR for logging instead
+
+Despite these limitations, Javy provides sufficient capabilities for writing effective Kubewarden validation policies through the host capabilities system.
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:53]:
```suggestion
Despite these limitations, Javy provides sufficient capabilities for writing effective Kubewarden validation policies through the hosts capabilities system.
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+
+## Tooling
+
+Writing Kubewarden policies requires:
+
+- **Node.js**: Version 18 or higher
+- **npm**: For dependency management
+- **TypeScript**: Recommended for type safety (optional)
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:61]:
```suggestion
- **Node.js**: Version 18 or higher.
- **npm**: For dependency management.
- **TypeScript**: Recommended for type safety (optional).
```

File: docs/tutorials/writing-policies/typescript/01-intro-typescript.md [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden
+```
+
+## Tutorial prerequisites
+
+During this tutorial you need these tools on your development machine:
+
+- **Node.js**: Version 18 or higher with npm for dependency management
+- [**`bats`**](https://github.com/bats-core/bats-core): Used to write the tests and automate their execution
+- [**`kwctl`**](https://github.com/kubewarden/kwctl/releases): CLI tool provided by Kubewarden to run its policies outside of Kubernetes, among other actions. It's covered in [the testing policies section](../../testing-policies/index.md) of the documentation.
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/01-intro-typescript.md:94]:
```suggestion
- **Node.js**: Version 18 or higher with npm for dependency management.
- [**`bats`**](https://github.com/bats-core/bats-core): Used to write the tests and automate their execution.
- [**`kwctl`**](https://github.com/kubewarden/kwctl/releases): CLI tool provided by Kubewarden to run its policies outside of Kubernetes, among other actions. It's covered in [the testing policies section](../../testing-policies/index.md) of the documentation.
```

File: docs/tutorials/writing-policies/typescript/02-scaffold.md [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+metadata:
+  name: nginx
+spec:
+  hostname: bad-host
+  containers:
+    - name: nginx
+      image: nginx:latest
+```
+
+But, it accepts the creation of this Pod:
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/02-scaffold.md:43]:
```suggestion
However, it accepts the creation of this Pod:
```

File: docs/tutorials/writing-policies/typescript/03-policy-settings.md [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/policy-settings"/>
+</head>
+
+:::danger Critical: Don't write logging information to STDOUT
+
+Writing to STDOUT breaks your policy. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.
+
+:::
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/03-policy-settings.md:20]:
```suggestion
:::danger Critical: Don't write logging information to STDOUT.

Writing to STDOUT breaks policies. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.

:::
```

File: docs/tutorials/writing-policies/typescript/04-validation.md [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/validation"/>
+</head>
+
+:::danger Critical: Don't write logging information to STDOUT
+
+Writing to STDOUT breaks your policy. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.
+
+:::
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/04-validation.md:20]:
```suggestion
:::danger Critical: Don't write logging information to STDOUT.

Writing to STDOUT breaks policies. Instead, use STDERR for logging or the logging facility provided by the Kubewarden SDK. The policy's output to STDOUT must only contain the validation response.

:::
```

File: docs/tutorials/writing-policies/typescript/04-validation.md [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+  case 'validate-settings':
+    validateSettings();
+    break;
+  default:
+    console.error('Unknown action:', action);
+    writeOutput(new Validation.Validation.ValidationResponse(false, 'Unknown policy action'));
+}
+```
+
+The Javy plugin provides the `policyAction()` function indicating whether the policy should validate a resource or validate its settings.
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/04-validation.md:174]:
```suggestion
The Javy plug-in provides the `policyAction()` function indicating whether the policy should validate a resource or validate its settings.
```

File: docs/tutorials/writing-policies/typescript/04-validation.md [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+- Get the relevant information from the incoming validation request.
+- Return a response based on the input and the policy settings.
+
+The incoming request contains a JSON object with the Kubernetes resource to validate.
+You can access this JSON object data through the Kubewarden SDK's helper functions.
+
+## The `validate` function
+
+The policy provided by the template already has a `validate` function in `src/index.ts`. 
+You can make use of it there, adding your logic to reject Pods with denied hostnames.
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/04-validation.md:35]:
```suggestion
The policy provided by the template already has a `validate` function in `src/index.ts`.
You can use it there, adding your logic to reject Pods with denied hostnames.
```

File: docs/tutorials/writing-policies/typescript/04-validation.md [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    return requestObject as KubernetesResource;
+  } catch (error) {
+    console.error('Error parsing Kubernetes resource:', error);
+    return undefined;
+  }
+}
+```
+
+This function handles the case where the Kubernetes object is provided as a JSON string or as an already parsed object.
```
Comment (jhkrug) [inline docs/tutorials/writing-policies/typescript/04-validation.md:135]:
```suggestion
This function handles the case where the Kubernetes object is a JSON string or an already parsed object.
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
Hi @esosaoh. I think this is the last set of changes. Just all for consistency and taking note of some terminology suggestions from the checkers. Nearly there. Thanks!

File: docs/tutorials/writing-policies/typescript/02-scaffold.md [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+  }
+}
+```
+
+Make sure to use a repository path that matches your actual GitHub repository.
+
+## Testing
+
+Provided the necessary tools are in place, the `make all` command builds the `annotated-policy.wasm` target. The command `make e2e` runs tests using `bats` with `kwctl`.
+
```
[the remark this answers is not in the document]
Comment (viccuad) [reply docs/tutorials/writing-policies/typescript/02-scaffold.md:80]: @esosaoh can you please check and list the needed changes for this to work?

[the remark this answers is not in the document]
Comment (esosaoh) [reply docs/tutorials/writing-policies/typescript/02-scaffold.md:80]: The Javy team has provided documentation with a steps to migrate to v2.0.0 of the `javy-plugin-api`: https://crates.io/crates/javy-plugin-api

**Review by flavio — COMMENTED** [review]
LGTM, I've left some minor comments.

I think we have to wait for kwctl 1.30 to be out before publishing this doc

**Comment by jhkrug** [comment]
Noting that the documented examples do not work, presumably as using kwctl 1.29 and waiting for 1.30. Or something else?:

```
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘ pwd
/home/jhk/tmp/kw-ts-test/js-policy-template
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘ npm install kubewarden-policy-sdk

up to date, audited 400 packages in 983ms

58 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘ ls -l node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden.wasm
-rw-r--r-- 1 jhk jhk 1146510 Oct 15 13:59 node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden.wasm
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘ cat package.json
{
	"name": "kubewarden-js-policy-template",
	"version": "0.1.0",
	"description": "jhk docs test",
	"main": "src/index.js",
	"scripts": {
		"build": "npx webpack --config webpack.config.cjs",
		"test": "bats e2e.bats",
		"format": "prettier --write .",
		"format:check": "prettier --check ."
	},
	"devDependencies": {
		"@types/jest": "^30.0.0",
		"@types/node": "^24.2.0",
		"jest": "^30.0.5",
		"kubernetes-types": "^1.30.0",
		"prettier": "^3.6.2",
		"ts-jest": "^29.4.1",
		"ts-loader": "^9.5.2",
		"typescript": "^5.8.3",
		"webpack": "^5.x.x",
		"webpack-cli": "^5.1.4"
	},
	"keywords": [
		"kubewarden",
		"policy",
		"kubernetes",
		"wasm",
		"javascript"
	],
	"author": "Kubewarden",
	"repository": {
		"type": "git",
		"url": "https://github.com/jhkrug/js-policy-template.git"
	},
	"bugs": {
		"url": "https://github.com/jhkrug/js-policy-template/issues"
	},
	"dependencies": {
		"kubewarden-policy-sdk": "^0.1.1"
	}
}
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
‽ make all
npm install

up to date, audited 400 packages in 833ms

58 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm run build

> kubewarden-js-policy-template@0.1.0 build
> npx webpack --config webpack.config.cjs

asset bundled.js 5.52 KiB [compared for emit] [minimized] (name: main)
asset types.d.ts 430 bytes [compared for emit]
asset index.d.ts 11 bytes [compared for emit]
./src/index.ts 3.84 KiB [built] [code generated]
./node_modules/kubewarden-policy-sdk/dist/bundle.js 3.85 KiB [built] [code generated]
webpack 5.101.3 compiled successfully in 1420 ms
npm install

up to date, audited 400 packages in 979ms

58 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
javy build dist/bundled.js -C plugin=node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden.wasm -o policy.wasm
Error: Could not process plugin: Using unsupported legacy plugin API
make: *** [Makefile:16: policy.wasm] Error 1
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘ make e2e
npm install

up to date, audited 400 packages in 840ms

58 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm run build

> kubewarden-js-policy-template@0.1.0 build
> npx webpack --config webpack.config.cjs

asset bundled.js 5.52 KiB [compared for emit] [minimized] (name: main)
asset types.d.ts 430 bytes [compared for emit]
asset index.d.ts 11 bytes [compared for emit]
./src/index.ts 3.84 KiB [built] [code generated]
./node_modules/kubewarden-policy-sdk/dist/bundle.js 3.85 KiB [built] [code generated]
webpack 5.101.3 compiled successfully in 1361 ms
javy build dist/bundled.js -C plugin=node_modules/kubewarden-policy-sdk/plugin/javy-plugin-kubewarden.wasm -o policy.wasm
Error: Could not process plugin: Using unsupported legacy plugin API
make: *** [Makefile:16: policy.wasm] Error 1
[~/tmp/kw-ts-test/js-policy-template][main *|u=]
⸘
```

**Comment by flavio** [comment]
Moving back to "to do" - it would be nice to test it out with kwctl 1.30.0-rc1 and address the issues found by @jhkrug 

We can look into that if @esosaoh doesn't have time

**Comment by esosaoh** [comment]
@flavio I'll look into this tomorrow

**Review by viccuad — APPROVED** [review]

**Review by jhkrug — CHANGES_REQUESTED** [review]
Ok, I got this working, but I think a couple of points need to be clearer as per the comments. wdyt? Thanks.
